Copy files provided by WKWebView clients on iOS that implement runOpenPanel instead of moving them

lilyspiniolas opened the pull request:
#### 05f1d0ebb69cc7cb3d8d9d9afd5e51d22f801e48
<pre>
Copy files provided by WKWebView clients on iOS that implement runOpenPanel instead of moving them
<a href="https://bugs.webkit.org/show_bug.cgi?id=280847">https://bugs.webkit.org/show_bug.cgi?id=280847</a>
<a href="https://rdar.apple.com/137172156">rdar://137172156</a>

Reviewed by Wenson Hsieh.

UIDelegate::UIClient::runOpenPanel now copies the URLs provided by implementations of runOpenPanel
in WKUIDelegates rather than moving them. This ensures that the workaround for
<a href="https://bugs.webkit.org/show_bug.cgi?id=228683">https://bugs.webkit.org/show_bug.cgi?id=228683</a> will not result in files being lost if the URL
points to an original file rather than an imported copy.

TestWebKitAPI.WebKit.RunOpenPanelNonLatin1 has also been updated to now create a file at
the URL it tests since file uploads from runOpenPanel on iOS now only proceed if the file
was successfully copied, which is impossible without an actual file.

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::runOpenPanel):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
(-[RunOpenPanelUIDelegate webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/284904@main">https://commits.webkit.org/284904@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5918336107886f7bfd610fe5f234788616b37632

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70527 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49933 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23292 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74616 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21705 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72644 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57733 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21545 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55881 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14348 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73593 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45409 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60800 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36342 "Found 1 new API test failure: /WPE/TestResources:/webkit/WebKitWebView/sync-request-on-max-conns (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42068 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18235 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20066 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64001 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18590 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76335 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14753 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17814 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63604 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14796 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60867 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63539 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11582 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5224 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10864 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45735 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/502 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46809 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48086 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46551 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
```diff
@@ -48,6 +48,7 @@ - (void)webView:(WKWebView *)webView runOpenPanelWithParameters:(WKOpenPanelPara
     EXPECT_FALSE(parameters.allowsDirectories);
     EXPECT_EQ(0ull, parameters._acceptedMIMETypes.count);
     EXPECT_EQ(0ull, parameters._acceptedFileExtensions.count);
+    [[NSFileManager defaultManager] createFileAtPath:[NSTemporaryDirectory() stringByAppendingPathComponent:expectedFileName] contents:nil attributes:nil];
```
Maybe `removeItemAtPath` before the end of the test?

lilyspiniolas replied:
Done!

cdumez commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
```diff
@@ -53,7 +53,13 @@ enum class PickerDismissalReason : uint8_t;
 
-+ (std::pair<RetainPtr<NSURL>, RetainPtr<NSURL>>)_copyToNewTemporaryDirectory:(NSURL *)originalURL fileCoordinator:(NSFileCoordinator *)fileCoordinator fileManager:(NSFileManager *)fileManager;
+struct TemporaryFileMoveResults {
+    bool success;
```
The initialization sites look a bit confusing. I think we should consider using a enum class for clarity. For example:
```
enum class WasSuccessful : bool { No, Yes };
```

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
```diff
@@ -53,7 +53,15 @@ enum class PickerDismissalReason : uint8_t;
 - (NSArray<NSString *> *)acceptedTypeIdentifiers;
 
-+ (std::pair<RetainPtr<NSURL>, RetainPtr<NSURL>>)_copyToNewTemporaryDirectory:(NSURL *)originalURL fileCoordinator:(NSFileCoordinator *)fileCoordinator fileManager:(NSFileManager *)fileManager;
+enum class MovedSuccessfully : bool { No, Yes };
```
Nit - let's put this new struct and enum class inside of the `WebKit` namespace, so that we won't collide with other symbols in the global namespace in the future.

lilyspiniolas committed:
Copy files provided by WKWebView clients on iOS that implement runOpenPanel instead of moving them
https://bugs.webkit.org/show_bug.cgi?id=280847
rdar://137172156

Reviewed by Wenson Hsieh.

UIDelegate::UIClient::runOpenPanel now copies the URLs provided by implementations of runOpenPanel
in WKUIDelegates rather than moving them. This ensures that the workaround for
https://bugs.webkit.org/show_bug.cgi?id=228683 will not result in files being lost if the URL
points to an original file rather than an imported copy.

TestWebKitAPI.WebKit.RunOpenPanelNonLatin1 has also been updated to now create a file at
the URL it tests since file uploads from runOpenPanel on iOS now only proceed if the file
was successfully copied, which is impossible without an actual file.

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::runOpenPanel):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.h:
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/RunOpenPanel.mm:
(-[RunOpenPanelUIDelegate webView:runOpenPanelWithParameters:initiatedByFrame:completionHandler:]):

Canonical link: https://commits.webkit.org/284904@main
